header/sync/sync.go

```diff
@@ -157,100 +142,37 @@ func (s *Syncer) wantSync() {
-	if err != nil {
-		log.Errorw("getting trusted head", "err", err)
+	newHead := s.pending.Head()
+	if newHead == nil {
 		return
 	}
 
-	s.syncTo(ctx, trstHead)
-}
-
-// processIncoming processes new processIncoming Headers, validates them and stores/caches if applicable.
-func (s *Syncer) processIncoming(ctx context.Context, maybeHead *header.ExtendedHeader) pubsub.ValidationResult {
-	// 1. Try to append. If header is not adjacent/from future - try it for pending cache below
-	_, err := s.store.Append(ctx, maybeHead)
-	switch err {
-	case nil:
-		// a happy case where we append adjacent header correctly
-		return pubsub.ValidationAccept
-	case header.ErrNonAdjacent:
-		// not adjacent, so try to cache it after verifying
-	default:
-		var verErr *header.VerifyError
-		if errors.As(err, &verErr) {
-			return pubsub.ValidationReject
-		}
-
-		log.Errorw("appending header",
-			"height", maybeHead.Height,
-			"hash", maybeHead.Hash().String(),
-			"err", err)
-		// might be a storage error or something else, but we can still try to continue processing 'maybeHead'
-	}
-
-	// 2. Get known trusted head, so we can verify maybeHead
-	trstHead, err := s.trustedHead(ctx)
-	if err != nil {
-		log.Errorw("getting trusted head", "err", err)
-		return pubsub.ValidationIgnore // we don't know if header is invalid so ignore
-	}
-
-	// 3. Filter out maybeHead if behind trusted
-	if maybeHead.Height <= trstHead.Height {
-		log.Warnw("received known header",
-			"height", maybeHead.Height,
-			"hash", maybeHead.Hash())
-		return pubsub.ValidationIgnore // we don't know if header is invalid so ignore
-	}
-
-	// 4. Verify maybeHead against trusted
-	err = trstHead.VerifyNonAdjacent(maybeHead)
-	var verErr *header.VerifyError
-	if errors.As(err, &verErr) {
-		log.Errorw("invalid header",
-			"height_of_invalid", maybeHead.Height,
-			"hash_of_invalid", maybeHead.Hash(),
-			"height_of_trusted", trstHead.Height,
-			"hash_of_trusted", trstHead.Hash(),
-			"reason", verErr.Reason)
-		return pubsub.ValidationReject
-	}
-
-	// 5. Save verified header to pending cache
-	// NOTE: Pending cache can't be DOSed as we verify above each header against a trusted one.
-	s.pending.Add(maybeHead)
-	// and trigger sync to catch-up
-	s.wantSync()
-	log.Infow("pending head",
-		"height", maybeHead.Height,
-		"hash", maybeHead.Hash())
-	return pubsub.ValidationAccept
-}
-
-// syncTo requests headers from locally stored head up to the new head.
-func (s *Syncer) syncTo(ctx context.Context, newHead *header.ExtendedHeader) {
 	head, err := s.store.Head(ctx)
 	if err != nil {
 		log.Errorw("getting head during sync", "err", err)
 		return
 	}
 
-	if head.Height == newHead.Height {
-		return
+	if head.Height >= newHead.Height {
+		log.Warnw("sync attempt to an already synced header",
+			"synced_height", head.Height,
+			"attempted_height", newHead.Height,
+		)
+		log.Warn("PLEASE REPORT THIS AS A BUG")
+		return // should never happen, but just in case
```

Comment:
If this should really never happen, shouldn't this be logged out as an error instead?

Reply:
Don't think so. This case does not break syncer